File: learning-core/executable/validation/python/test_validation.py
import pytest

from validation import Registration, parse_registration


def test_valid_payload_is_normalized() -> None:
    assert parse_registration({"username": "  Ada  ", "age": 36}) == Registration("Ada", 36)


@pytest.mark.parametrize(
    "payload",
    [
        None,
        [],
        {},
        {"username": "Ada", "age": True},
        {"username": "", "age": 36},
        {"username": "Ada", "age": -1},
        {"username": "Ada", "age": 131},
        {"username": "Ada", "age": 36, "admin": True},
    ],
)
def test_invalid_payloads_are_rejected(payload: object) -> None:
    with pytest.raises(ValueError):
        parse_registration(payload)

File: learning-core/executable/validation/cpp/test_validation.cpp
#include <cassert>
#include <stdexcept>
#include <string>

#include "validation.cpp"

static bool rejects(const std::string& name, int age) {
    try {
        (void)parse_registration(name, age);
        return false;
    } catch (const std::invalid_argument&) {
        return true;
    }
}

int main() {
    const auto valid = parse_registration("  Ada  ", 36);
    assert(valid.username == "Ada");
    assert(valid.age == 36);

    assert(rejects("", 36));
    assert(rejects("Ada", -1));
    assert(rejects("Ada", 131));
}
